src/datachain/query/dataset.py

```diff
@@ -618,43 +681,349 @@ def clone(self, partition_by: PartitionByType | None = None) -> "Self":
         if partition_by is not None:
             return self.__class__(
                 self.udf,
-                self.catalog,
+                self.session,
                 partition_by=partition_by,
                 parallel=self.parallel,
                 workers=self.workers,
                 min_task_size=self.min_task_size,
                 batch_size=self.batch_size,
             )
-        return self.__class__(self.udf, self.catalog)
+        return self.__class__(self.udf, self.session)
+
+    def _checkpoint_exist(self, _hash: str, partial: bool = False) -> Checkpoint | None:
+        """
+        Check if checkpoint exists for given hash.
+        Returns the Checkpoint object if found, None otherwise.
+        Checks current job first, then parent job if it exists.
+        """
+        checkpoints_reset = env2bool("DATACHAIN_CHECKPOINTS_RESET", undefined=False)
+
+        if (
+            self.job.parent_job_id
+            and not checkpoints_reset
+            and (
+                checkpoint := self.metastore.find_checkpoint(
+                    self.job.parent_job_id, _hash, partial=partial
+                )
+            )
+        ):
+            return checkpoint
+
+        return None
+
+    @property
+    def job(self) -> Job:
+        return self.session.get_or_create_job()
+
+    @property
+    def metastore(self):
+        return self.session.catalog.metastore
+
+    @property
+    def warehouse(self):
+        return self.session.catalog.warehouse
+
+    @staticmethod
+    def input_table_name(job_id: str, _hash: str) -> str:
+        """Job-specific input table name."""
+        return f"udf_{job_id}_{_hash}_input"
+
+    @staticmethod
+    def output_table_name(job_id: str, _hash: str) -> str:
+        """Job-specific final output table name."""
+        return f"udf_{job_id}_{_hash}_output"
+
+    @staticmethod
+    def partial_output_table_name(job_id: str, _hash: str) -> str:
+        """Job-specific partial output table name."""
+        return f"udf_{job_id}_{_hash}_output_partial"
+
+    def get_or_create_input_table(self, query: Select, _hash: str) -> "Table":
+        """
+        Get or create input table for the given hash.
+
+        First checks if current job has the input table.
+        If not, searches ancestor jobs and uses their table directly.
+        If not found in any ancestor, creates it for current job from query.
+
+        Returns the input table (may belong to current job or an ancestor).
+        """
+        current_input_table_name = UDFStep.input_table_name(self.job.id, _hash)
+
+        # Check if current job already has the input table
+        if self.warehouse.db.has_table(current_input_table_name):
+            return self.warehouse.get_table(current_input_table_name)
+
+        # Search ancestor jobs for the input table
+        if self.job.parent_job_id:
+            ancestor_job_ids = self.metastore.get_ancestor_job_ids(self.job.id)
```

Comment:
Why do we need to do recursive search? why single parent not enough?

Reply:
okay, I think I understand - bc we are reusing it ... 

but it seems too complicated ... all this recursive search ...

so, we have hash  ... we have a JOB that we run a few times. Can we just use some job hash + chain hash or udf hash? Instead of using unique job_id + hash. Or even just a single hash to search for a saved table?

Is there a case that there will two UDF tables for the same query (_hash) for the same job that we run multiple times?